### PR TITLE
Use rapidjson in loki

### PR DESCRIPTION
**Pull request by xlqian**

### CODE REVIEW

File: src/loki/isochrone_action.cc [hunk 1]
```diff
@@ -40,38 +43,42 @@ namespace valhalla {
...
       init_isochrones(request);
       //check that location size does not exceed max
       if (locations.size() > max_locations.find("isochrone")->second)
         throw valhalla_exception_t{400, 150, std::to_string(max_locations.find("isochrone")->second)};
 
-      auto costing = request.get<std::string>("costing");
-      auto date_type = request.get_optional<int>("date_time.type");
+      auto costing = GetOptionalFromRapidJson<std::string>(request, "/costing");
+      if (! costing)
+        costing = "";
```
Comment (TeXitoi) [inline src/loki/isochrone_action.cc]: if auto pattern

File: src/loki/isochrone_action.cc [hunk 2]
```diff
@@ -83,18 +90,20 @@ namespace valhalla {
         //correlate the various locations to the underlying graph
         const auto projections = loki::Search(locations, reader, edge_filter, node_filter);
         for(size_t i = 0; i < locations.size(); ++i) {
-          request.put_child("correlated_" + std::to_string(i), projections.at(locations[i]).ToPtree(i));
+          auto path_tmp = ("/correlated_" + std::to_string(i)).c_str();
```
Comment (TeXitoi) [inline src/loki/isochrone_action.cc]: use after free (temporary `std::string` is deleted after this line, but path_tmp point to it)

File: src/loki/locate_action.cc [hunk 3]
```diff
@@ -123,38 +126,38 @@ namespace {
...
     }
 
-    worker_t::result_t loki_worker_t::locate(const boost::property_tree::ptree& request, http_request_info_t& request_info) {
+    worker_t::result_t loki_worker_t::locate(const rapidjson::Document& request, http_request_info_t& request_info) {
       init_locate(request);
       //correlate the various locations to the underlying graph
       auto json = json::array({});
-      auto verbose = request.get<bool>("verbose", false);
+      auto verbose_optinal = GetOptionalFromRapidJson<bool>(request, "/verbose");
+      auto verbose = verbose_optinal ? *verbose_optinal : false;
```
Comment (TeXitoi) [inline src/loki/locate_action.cc]:
```c++
const verbose = GetOptionalFromRapidJson<bool>(request, "/verbose").value_or(false);

File: src/loki/matrix_action.cc [hunk 4]
```diff
@@ -61,10 +63,12 @@ namespace {
 namespace valhalla {
   namespace loki {
 
-    void loki_worker_t::init_matrix(ACTION_TYPE action, boost::property_tree::ptree& request) {
-      auto request_locations = request.get_child_optional("locations");
-      auto request_sources = request.get_child_optional("sources");
-      auto request_targets = request.get_child_optional("targets");
+    void loki_worker_t::init_matrix(ACTION_TYPE action, rapidjson::Document& request) {
+      //auto* request_locations = rapidjson::Pointer("/locations").Get(request);
```
Comment (TeXitoi) [inline src/loki/matrix_action.cc]: useless comment

File: src/loki/matrix_action.cc [hunk 5]
```diff
@@ -170,8 +176,9 @@ namespace valhalla {
         for(size_t i = 0; i < sources_targets.size(); ++i) {
           const auto& l = sources_targets[i];
           const auto& projection = searched.at(l);
-          request.put_child("correlated_" + std::to_string(i), projection.ToPtree(i));
-
+          std::string name_tmp = "/correlated_" + std::to_string(i);
+          //rapidjson::SetValueByPointer(request, name_tmp.c_str(), correlated.ToRapidJson(i, request.GetAllocator()));
```
Comment (TeXitoi) [inline src/loki/matrix_action.cc]: useless comment

File: src/loki/route_action.cc [hunk 6]
```diff
@@ -37,26 +41,28 @@ namespace {
...
-      if (costing == "multimodal" || costing == "transit") {
-        auto transit_start_end_max_distance =
-            request.get<int>("costing_options.pedestrian.transit_start_end_max_distance", min_transit_walking_dis);
-        auto transit_transfer_max_distance =
-            request.get<int>("costing_options.pedestrian.transit_transfer_max_distance", min_transit_walking_dis);
+      if (*costing == "multimodal" || *costing == "transit") {
+        auto transit_start_end_max_distance = rapidjson::GetValueByPointerWithDefault(request,
+            "/costing_options/pedestrian/transit_start_end_max_distance", min_transit_walking_dis).GetUint();
+        auto transit_transfer_max_distance =rapidjson::GetValueByPointerWithDefault(request,
+            "/costing_options/pedestrian/transit_transfer_max_distance", min_transit_walking_dis).GetUint();
```
Comment (TeXitoi) [inline src/loki/route_action.cc]: Is it working if `/costing_options/pedestrian/transit_transfer_max_distance` is not an integer?

File: src/loki/route_action.cc [hunk 7]
```diff
@@ -67,39 +73,38 @@ namespace valhalla {
...
+      if (! date_type_pointer.Get(request) && (*costing == "multimodal" || *costing == "transit")) {
+        date_type_pointer.Set(request, 0);
       }
-
+      auto date_type = GetOptionalFromRapidJson<int>(request, "/date_time/type");
+      auto& locations_array = request["/locations"];
       //check the date stuff
-      auto date_time_value = request.get_optional<std::string>("date_time.value");
+      auto date_time_value = GetOptionalFromRapidJson<std::string>(request, "/date_time/value");
       if (date_type) {
```
Comment (TeXitoi) [inline src/loki/route_action.cc]: if auto pattern

File: src/loki/route_action.cc [hunk 8]
```diff
@@ -113,7 +118,9 @@ namespace valhalla {
         const auto projections = loki::Search(locations, reader, edge_filter, node_filter);
         for(size_t i = 0; i < locations.size(); ++i) {
           const auto& correlated = projections.at(locations[i]);
-          request.put_child("correlated_" + std::to_string(i), correlated.ToPtree(i));
+          std::string name_tmp = "/correlated_" + std::to_string(i);
+          //rapidjson::SetValueByPointer(request, name_tmp.c_str(), correlated.ToRapidJson(i, request.GetAllocator()));
```
Comment (TeXitoi) [inline src/loki/route_action.cc]: useless comment

File: src/loki/service.cc [hunk 9]
```diff
@@ -45,25 +48,27 @@ namespace {
...
-      }//no json parameter, check the body
-      else if(!request.body.empty()) {
-        std::istringstream is(request.body);
-        boost::property_tree::read_json(is, pt);
-      }
+    const auto& json = request.query.find("json");
+    if (json != request.query.end() && json->second.size()
+      && json->second.front().size()) {
+      std::istringstream is(json->second.front());
+      d.Parse(is.str().c_str());
```
Comment (TeXitoi) [inline src/loki/service.cc]: `stringstream` not needed for RapidJson

File: src/loki/service.cc [hunk 10]
```diff
@@ -45,25 +48,27 @@ namespace {
...
+    }//no json parameter, check the body
+    else if(!request.body.empty()) {
+      std::istringstream is(request.body);
+      d.Parse(is.str().c_str());
```
Comment (TeXitoi) [inline src/loki/service.cc]: `stringstream` not needed for rapidjson

File: src/loki/service.cc [hunk 11]
```diff
@@ -74,34 +79,33 @@ namespace {
 
       //turn single value entries into single key value
       if(kv.second.size() == 1) {
-        pt.add(kv.first, kv.second.front());
+        d.AddMember(rapidjson::StringRef(kv.second.front().c_str()), rapidjson::StringRef(kv.first.c_str()), allocator);
```
Comment (TeXitoi) [inline src/loki/service.cc]: Shure there is no lifetime problem?

File: src/loki/service.cc [hunk 12]
```diff
@@ -74,34 +79,33 @@ namespace {
...
-      auto& array = pt.put_child("locations", boost::property_tree::ptree());
-      for(const auto& location : pt.get_child("loc")) {
-        Location l = Location::FromCsv(location.second.get_value<std::string>());
-        boost::property_tree::ptree element;
-        element.put("lon", l.latlng_.first);
-        element.put("lat", l.latlng_.second);
-        array.push_back(std::make_pair("", element));
+      auto& array = rapidjson::Pointer("/locations").Set(d, rapidjson::Value{rapidjson::kArrayType});
+
+      for(const auto& location : d["loc"].GetArray()) {
```
Comment (TeXitoi) [inline src/loki/service.cc]: It can't fail?

File: src/loki/service.cc [hunk 13]
```diff
@@ -132,27 +136,27 @@ namespace valhalla {
...
-          locations.push_back(baldr::Location::FromPtree(location.second));
+          locations.push_back(baldr::Location::FromRapidJson(location));
         }
         catch (...) {
           throw valhalla_exception_t{400, 130};
         }
       }
       if (!healthcheck)
-        valhalla::midgard::logging::Log("location_count::" + std::to_string(request_locations->size()), " [ANALYTICS] ");
+        valhalla::midgard::logging::Log("location_count::" + std::to_string(request["locations"].Size()), " [ANALYTICS] ");
```
Comment (TeXitoi) [inline src/loki/service.cc]:
```c++
locations_array->Size()
```

File: src/loki/service.cc [hunk 14]
```diff
@@ -247,12 +255,13 @@ namespace valhalla {
...
-        jsonp = request_pt.get_optional<std::string>("jsonp");
+        auto request_rj = from_request(action->second, request);
+        jsonp = GetOptionalFromRapidJson<std::string>(request_rj, "/jsonp");
         //let further processes more easily know what kind of request it was
-        request_pt.put<int>("action", action->second);
+        rapidjson::SetValueByPointer(request_rj, "/action", action->second);
         //flag healthcheck requests; do not send to logstash
-        healthcheck = request_pt.get<bool>("healthcheck", false);
+        auto optional_healthcheck = GetOptionalFromRapidJson<bool>(request_rj, "/healthcheck");
+        healthcheck = optional_healthcheck ? *optional_healthcheck : false;
```
Comment (TeXitoi) [inline src/loki/service.cc]:
```c++
const auto healthcheck = GetOptionalFromRapidJson<bool>(request_rj, "/healthcheck").value_or(false);
```

File: src/loki/route_action.cc (hunk 6 above)
Comment (xlqian) [inline src/loki/route_action.cc]: if fails..

File: src/loki/route_action.cc (hunk 7 above)
Comment (xlqian) [inline src/loki/route_action.cc]: done

File: src/loki/service.cc (hunk 9 above)
Comment (xlqian) [inline src/loki/service.cc]: done

File: src/loki/service.cc (hunk 10 above)
Comment (xlqian) [inline src/loki/service.cc]: done

File: src/loki/service.cc (hunk 11 above)
Comment (xlqian) [inline src/loki/service.cc]: i change that to a copy

File: src/loki/service.cc (hunk 12 above)
Comment (xlqian) [inline src/loki/service.cc]: yes

**Pull request closed without merging**